gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -154,76 +164,120 @@ private void createEmptyWorkUnitsForSkippedPartitions(Map<String, List<WorkUnit>
   }
 
   /**
-   * Group workUnits into multiWorkUnits. Each input workUnit corresponds to a (topic, partition).
+   * Group {@link WorkUnit}s into {@link MultiWorkUnit}s. Each input {@link WorkUnit} corresponds to
+   * a (topic, partition).
    *
-   * There are two levels of grouping. In the first level, some workunits corresponding to partitions
-   * of the same topic are grouped together into a single workunit. This reduces the number of small
+   * There are two levels of grouping. In the first level, some {@link WorkUnit}s corresponding to partitions
+   * of the same topic are grouped together into a single {@link WorkUnit}. This reduces the number of small
    * output files since these partitions will share output files, rather than each creating individual
-   * output files. The number of grouped workunits is approximately 3 * numOfMultiWorkunits, since the
-   * worst-fit-decreasing algorithm (used by the second level) should work well if the number of items
+   * output files. The number of grouped {@link WorkUnit}s is approximately 3 * number of {@link MultiWorkUnit}s,
+   * since the worst-fit-decreasing algorithm (used by the second level) should work well if the number of items
    * is more than 3 times the number of bins.
    *
-   * In the second level, these grouped workunits are assembled into multiWorkunits using worst-fit-decreasing.
+   * In the second level, these grouped {@link WorkUnit}s are assembled into {@link MultiWorkunit}s
+   * using worst-fit-decreasing.
    *
-   * @param workUnits A two-dimensional list of workunits, each corresponding to a single Kafka partition, grouped
-   * by topics.
-   * @param numOfMultiWorkunits Desired number of MultiWorkUnits.
+   * @param workUnitsByTopic A two-dimensional list of {@link WorkUnit}s, each corresponding to a single Kafka partition,
+   * grouped by topics.
+   * @param numOfMultiWorkunits Desired number of {@link MultiWorkUnit}s.
    * @param state
-   * @return A list of MultiWorkUnits.
+   * @return A list of {@link MultiWorkUnit}s.
    */
-  private List<WorkUnit> getMultiWorkunits(Map<String, List<WorkUnit>> workUnits, int numOfMultiWorkunits,
+  private List<WorkUnit> getMultiWorkunits(Map<String, List<WorkUnit>> workUnitsByTopic, int numOfMultiWorkunits,
       SourceState state) {
     Preconditions.checkArgument(numOfMultiWorkunits >= 1);
 
-    long totalEstDataSize = 0;
-    for (List<WorkUnit> workUnitsForTopic : workUnits.values()) {
+    double totalEstDataSize = 0;
+    for (List<WorkUnit> workUnitsForTopic : workUnitsByTopic.values()) {
       for (WorkUnit workUnit : workUnitsForTopic) {
         setWorkUnitEstSize(workUnit);
         totalEstDataSize += getWorkUnitEstSize(workUnit);
       }
     }
-    long avgGroupSize = (long) ((double) totalEstDataSize / (double) numOfMultiWorkunits / 3.0);
+    if (isWorkUnitPreGroupingEnabled(state)) {
```

Comment:
Can you abstract-away the bin-packing strategy with an interface and a couple of implementing classes. This will make adding new strategies easier in the future and the code cleaner?